doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,396 @@ Applying theorems
+
+   :n:`@one_term` (inside :n:`@one_term_with_bindings`)
+     If :n:`@one_term` is an :n:`@ident`, it is the name of
+     a theorem, lemma or hypothesis whose type is given in the
+     theorem statement or shown in the context.  Otherwise it is a proof term whose
+     type can be displayed with :cmd:`Check` :n:`@one_term`.
+
+   Without :n:`@in_hyp_as` (the goal case)
+     If the goal matches all of the type of :n:`@one_term` (both premises and
+     the conclusion), the tactic proves the goal.
+     Otherwise, the tactic matches the goal against the conclusion of :n:`@one_term`
+     and, if necessary, one or more premises (from right to left).
+     If the match succeeds, the tactic replaces the current goal with a subgoal for
+     each non-dependent premise of the type of :n:`@one_term`.  This
+     :ref:`example <apply_backward>` matches only the conclusion, while
+     this :ref:`one <apply_backward_w_premises>` also matches a premise.
+
+     If the conclusion of the type of :token:`one_term` does
+     not match the goal *and* the conclusion is an inductive type isomorphic to
+     a tuple type, then each component of the tuple is recursively matched to
+     the goal in left-to-right order.
+
+     .. note::
+
+        When the conclusion of the type of the term to ``apply`` is an inductive
+        type isomorphic to a tuple type and ``apply`` looks recursively to check whether a
+        component of the tuple matches the goal, it excludes components whose
+        statement would result in applying an universal lemma of the form
+        ``forall A, … -> A``.
+
+     The goal case uses first-order unification with dependent types unless the
+     conclusion of the type of :token:`term` is of the form
+     :n:`P (t__1 … t__n)` with ``P`` to be instantiated. In the latter case,
+     the behavior depends on the form of the target. If the target is of the form
+     :n:`(fun x__1 … x__n => Q) u__1 … u__n` and the :n:`t__i` and :n:`u__i` unify,
+     then :g:`P` is instantiated into :n:`(fun x__1 … x__n => Q)`. Otherwise,
+     :tacn:`apply`
+     tries to define :g:`P` by abstracting over :n:`t__1 … t__n` in the target.
+     You can use :tacn:`pattern` to transform the target so that it
+     gets the form :n:`(fun x__1 … x__n => Q) u__1 … u__n`.  See the example
+     :ref:`here <example_apply_pattern>`.
+
+   :n:`@in_hyp_as` (the hypothesis case)
+     Proceeding from *right to left*, find the first premise of the type of
+     :n:`@one_term` that matches the specified hypothesis.  If a match
+     is found, the hypothesis is replaced with the conclusion of the type of
+     :n:`one_term` (substituting for the unified variables)
+     and the tactic creates a new subgoal for each unmatched premise.
+     See the example :ref:`here <apply_forward>`.
+
+     If specified, :n:`as @simple_intropattern` is applied to the conclusion
+     of the type of :n:`@one_term`.
```

Comment:
Maybe add:
```suggestion
     of the type of :n:`@one_term` (in this case, the selected hypothesis
     may be left unchanged if its name is not reused).
```

Reply:
"may be left unchanged" or "is left unchanged"?

Reply:
"is left unchanged"